File: solo/puzzles/easy/mars-lander-episode-1/mars_lander_episode_1.cpp
#include <iostream>
#include <string>

int main()
{
   int surface_n; // the number of points used to draw the surface of Mars.
   std::cin >> surface_n; std::cin.ignore();
   for (int i = 0; i < surface_n; i++)
   {
      int land_x; // X coordinate of a surface point. (0 to 6999)
      int land_y; // Y coordinate of a surface point. By linking all the points together in a sequential fashion, you form the surface of Mars.
      std::cin >> land_x >> land_y; std::cin.ignore();
   }

   // game loop
   while (1)
   {
      int x;
      int y;
      int h_speed; // the horizontal speed (in m/s), can be negative.
      int v_speed; // the vertical speed (in m/s), can be negative.
      int fuel; // the quantity of remaining fuel in liters.
      int rotate; // the rotation angle in degrees (-90 to 90).
      int power; // the thrust power (0 to 4).
      std::cin >> x >> y >> h_speed >> v_speed >> fuel >> rotate >> power; std::cin.ignore();

      if (y > 1500)
      {
         std::cout << "0 3" << std::endl;
      }
      else if (y > 200)
      {
         std::cout << "0 4" << std::endl;
      }
      else
      {
         std::cout << "0 0" << std::endl;
      }
   }
}

File: solo/puzzles/easy/mars-lander-episode-1/mars_lander_episode_1.rs
use std::io;

macro_rules! parse_input {
   ($x:expr, $t:ident) => ($x.trim().parse::<$t>().unwrap())
}

fn main() {
   let mut input_line = String::new();
   io::stdin().read_line(&mut input_line).unwrap();
   let surface_n = parse_input!(input_line, i32); // the number of points used to draw the surface of Mars.
   for i in 0..surface_n as usize {
      let mut input_line = String::new();
      io::stdin().read_line(&mut input_line).unwrap();
      let inputs = input_line.split(" ").collect::<Vec<_>>();
      let land_x = parse_input!(inputs[0], i32); // X coordinate of a surface point. (0 to 6999)
      let land_y = parse_input!(inputs[1], i32); // Y coordinate of a surface point. By linking all the points together in a sequential fashion, you form the surface of Mars.
   }

   loop {
      let mut input_line = String::new();
      io::stdin().read_line(&mut input_line).unwrap();
      let inputs = input_line.split(" ").collect::<Vec<_>>();
      let x = parse_input!(inputs[0], i32);
      let y = parse_input!(inputs[1], i32);
      let h_speed = parse_input!(inputs[2], i32); // the horizontal speed (in m/s), can be negative.
      let v_speed = parse_input!(inputs[3], i32); // the vertical speed (in m/s), can be negative.
      let fuel = parse_input!(inputs[4], i32); // the quantity of remaining fuel in liters.
      let rotate = parse_input!(inputs[5], i32); // the rotation angle in degrees (-90 to 90).
      let power = parse_input!(inputs[6], i32); // the thrust power (0 to 4).

      if (y > 1500) {
         println!("0 3");
      } else if (y > 200) {
         println!("0 4");
      } else {
         println!("0 0");
      }
   }
}
